### PR TITLE
GPT-2 Fine-Tune Replication

### PR DESCRIPTION
**Pull request by luciaquirke**
- Add a learning rate scheduler config to bring the available schedules in line with HF (with the addition of lr_start in polynomial, required for the GPT-2 fine-tune).
- Add Trainer resume option for training phase (future work: backward phase)
- Drop data that doesn't fit evenly into the world size or batch size
- Add GPT-2 fine-tune replication yaml
- Fix a bug in distributed.py DTensor usage that pops up when models have tensors that aren't cleanly divisible across ranks (only applies to GPT-2 so far)

- Add validate_batch_size. This function may be useless because it does the same thing as the first training step. But we've had issues in the past that haven't been fully diagnosed where a too-large batch size in FSDP can cause a NCCL hang rather than an immediate crash. This can happen mid-training if the batch size is variable. If this happens it might be nice to have it happen in a dedicated location. I'm not sure if this will be relevant to us yet, need to look into our planned batching strategies.
 
Future work:
- Add fake data to make batches even without dropping data, and ignore their gradients

### CODE REVIEW

File: bergson/magic/cli.py [hunk 1]
```diff
@@ -96,7 +100,88 @@ def compute_query_gradients(
...
+                return _warmup(step)
+            return lr
+
+    elif scheduler_type == "linear":
+
+        def _schedule(step):
+            if step < warmup_steps:
+                return _warmup(step)
+            progress = (step - warmup_steps) / decay_steps
+            return lr - (lr - lr_end) * progress
```
Comment (davidoj) [inline bergson/magic/cli.py]: I don't know if it practically matters very much but if you ever call get_schedule with num_steps < the actual number of training steps you take this will behave oddly (why? Maybe you want to bodgy a three-phase scheduler or something). Could clamp `progress = max(1, (step - warmup_steps) / decay_steps)` to avoid this

**Review by davidoj — COMMENTED** [review]
Looks fine, left a couple of comments, address them at your discretion

**Review by norabelrose — CHANGES_REQUESTED** [review]

**Review by davidoj — APPROVED** [review]
Looks good to me
